### PR TITLE
Add components to the bundlectl CLI

### LINKED ISSUE #88: Add tests for bundleio (cli)
We should add tests for `bundleio.go` in `pkg/commands/cmdlib/bundleio.go`

### LINKED ISSUE #106: Handle ComponentPackages with the CLI
Right now `bundlectl` only supports bundles. It should support component packages too.

### PR DESCRIPTION
**Pull request by Kashomon**
This CLI adds component support to the Bundlectl CLI by changing bundle
file to input file and adding an input bundle type flag.

Fixes #106 
Fixes #88

### CODE REVIEW

**Comment by Kashomon** [comment]
Note: This PR definitely needs better unit tests, but I thought I'd send this out now since it's functional and it supersedes a previous PR: (#94)

File: pkg/commands/cmdlib/bundleio.go [hunk 1]
```diff
@@ -29,88 +29,218 @@ import (
...
+	return bw, nil
 }
 
-// WriteStructuredContents writes some structured contents from some object `obj`. The contents must
-// be serializable to both JSON and YAML.
-func WriteStructuredContents(ctx context.Context, obj interface{}, rw files.FileReaderWriter, g *GlobalOptions) error {
-	bytes, err := converter.FromObject(obj).ToContentType(g.OutputFormat)
+// WriteBundleData writes either the component or bundle object from the BundleWrapper.
+func (brw *BundleReaderWriter) WriteBundleData(ctx context.Context, bw *BundleWrapper, g *GlobalOptions) error {
+	if bw.Bundle != nil {
```
Comment (justinsb) [inline pkg/commands/cmdlib/bundleio.go]: As this function returns error maybe we could catch both fields being set here?

File: pkg/commands/cmdlib/bundleio.go [hunk 2]
```diff
@@ -29,88 +29,218 @@ import (
...
+		return brw.WriteStructuredContents(ctx, bw.Component, g)
+	} else {
+		return fmt.Errorf("both bundle and component were nil")
+	}
+}
+
+// WriteStructuredContents writes some structured contents from some object
+// `obj`. The contents must be serializable to both JSON and YAML.
+func (brw *BundleReaderWriter) WriteStructuredContents(ctx context.Context, obj interface{}, g *GlobalOptions) error {
+	outFmt := g.OutputFormat
```
Comment (justinsb) [inline pkg/commands/cmdlib/bundleio.go]: Is the file extension overriding the flag?

File: pkg/commands/cmdlib/bundleio_test.go [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+	rdBytes []byte
+	rdErr   error
+
+	wrPath  string
+	wrBytes []byte
+	wrError error
+}
+
+func (rw *fakeFileRW) ReadFile(_ context.Context, path string) ([]byte, error) {
+	rw.rdPath = path
```
Comment (justinsb) [inline pkg/commands/cmdlib/bundleio_test.go]: Nit: you could return an error here if we tried to read two files

File: pkg/commands/cmdlib/bundleio_test.go [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+	return rw.rdBytes, rw.rdErr
+}
+
+func (rw *fakeFileRW) WriteFile(_ context.Context, path string, bytes []byte, permissions os.FileMode) error {
+	rw.wrPath = path
```
Comment (justinsb) [inline pkg/commands/cmdlib/bundleio_test.go]: Similar here - an error if wrPath != "" ?

File: pkg/commands/cmdlib/global_options.go [hunk 5]
```diff
@@ -14,25 +14,37 @@
...
 
-	// Whether to inline the components in the bundle before doing any processing
+	// InlineComponents indicates whether to inline the components in the bundle
+	// before doing any processing. Only applies when the BundleFile option is
+	// specified.
 	InlineComponents bool
 
-	// Whether to inline tho objects before doing any processing (typically
-	// requires InlineComponents to be useful)
+	// InlineObjects indicates whether to inline tho objects before doing any
```
Comment (justinsb) [inline pkg/commands/cmdlib/global_options.go]: Nit: typo in `tho`

File: pkg/commands/root.go [hunk 6]
```diff
@@ -30,26 +30,24 @@ import (
 // AddCommands adds all subcommands to the root command.
 func AddCommands(ctx context.Context, args []string) *cobra.Command {
 	rootCmd := &cobra.Command{
-		Use:   "bundler",
-		Short: "bundler is tool for inspecting, validation, and modifying components packages and component sets.",
+		Use:   "bundlectl",
```
Comment (justinsb) [inline pkg/commands/root.go]: So k8s :-)

File: pkg/commands/root.go [hunk 7]
```diff
@@ -30,26 +30,24 @@ import (
...
 	rootCmd.PersistentFlags().StringVarP(
-		&cmdlib.GlobalOptionsValues.InputFormat, "in-format", "", "yaml", "The input file format. One of either 'json' or 'yaml'")
+		&cmdlib.GlobalOptionsValues.InputFormat, "in-format", "", "yaml", "The input file format. One of either 'json' or 'yaml'. If an input-file is specified, it is inferred from the file extension.")
 
 	rootCmd.PersistentFlags().StringVarP(
 		&cmdlib.GlobalOptionsValues.OutputFile, "output-file", "o", "", "The path for any output file")
 
 	rootCmd.PersistentFlags().StringVarP(
-		&cmdlib.GlobalOptionsValues.OutputFormat, "format", "", "yaml", "The output file format. One of either 'json' or 'yaml'")
+		&cmdlib.GlobalOptionsValues.OutputFormat, "format", "", "yaml", "The output file format. One of either 'json' or 'yaml'. If an output-file is specified, it is inferred from the file extension.")
```
Comment (justinsb) [inline pkg/commands/root.go]: Ah - I see the flag precedence issue now.  If it's important, the approach I normally use is either to make the default decision late (at the point of use), or alternatively to infer the OutputFormat from the filename asap.

**Comment by justinsb** [comment]
Generally looks good, but I don't really understand stdioReaderWriter vs io.ReadWriter (I think they're the same, but ...)

File: pkg/commands/cmdlib/bundleio.go [hunk 8]
```diff
@@ -29,88 +29,218 @@ import (
...
-// ReadBundle reads component data file contents from a file or stdin.
-func ReadBundle(ctx context.Context, rw files.FileReaderWriter, g *GlobalOptions) (*bundle.Bundle, error) {
+// BundleWrapper represents a collection of possibly specified bundle types.
+type BundleWrapper struct {
+	Bundle    *bundle.Bundle
+	Component *bundle.ComponentPackage
+}
+
+// AllComponents returns all the components in the BundleWrapper, with the
+// assumption that only one of Bundle or Component is filled out.
```
[the remark this answers is not in the document]
Comment (Kashomon) [reply pkg/commands/cmdlib/bundleio.go]: Yes I agree -- but I it's meant to be an internal type, only created by this method and used by the various command libraries

File: pkg/commands/cmdlib/bundleio.go [hunk 9]
```diff
@@ -29,88 +29,218 @@ import (
...
+	return inline.NewInlinerWithScheme(
+		files.FileScheme,
+		&files.LocalFileObjReader{filepath.Dir(inputFile), rw},
+		inline.DefaultPathRewriter,
+	)
+}
+
+// stdioReaderWriter is a source from stdin and sink to stdout.
+type stdioReaderWriter interface {
+	readBytes() ([]byte, error)
```
[the remark this answers is not in the document]
Comment (Kashomon) [reply pkg/commands/cmdlib/bundleio.go]: Oh yeah, that works

File: pkg/commands/cmdlib/bundleio.go [hunk 10]
```diff
@@ -29,88 +29,218 @@ import (
...
+	writeBytes([]byte) error
+}
+
+type realStdioReaderWriter struct{}
+
+func (r *realStdioReaderWriter) readBytes() ([]byte, error) {
+	var bytes []byte
```
[the remark this answers is not in the document]
Comment (Kashomon) [reply pkg/commands/cmdlib/bundleio.go]: Oh yeah, that works. \o/. Nice

File: pkg/commands/cmdlib/bundleio.go (hunk 1 above)
Comment (Kashomon) [reply pkg/commands/cmdlib/bundleio.go]: Done

File: pkg/commands/cmdlib/bundleio.go (hunk 2 above)
Comment (Kashomon) [reply pkg/commands/cmdlib/bundleio.go]: Yeah, it is. Updated based on your comment below.

File: pkg/commands/cmdlib/global_options.go (hunk 5 above)
Comment (Kashomon) [reply pkg/commands/cmdlib/global_options.go]: Fixed

File: pkg/commands/root.go (hunk 6 above)
Comment (Kashomon) [reply pkg/commands/root.go]: ;)

File: pkg/commands/root.go (hunk 7 above)
Comment (Kashomon) [reply pkg/commands/root.go]: Done. Went with default decision late.

File: pkg/commands/cmdlib/bundleio.go [hunk 11]
```diff
@@ -29,88 +29,218 @@ import (
...
+// BundleWrapper represents a collection of possibly specified bundle types.
+type BundleWrapper struct {
+	Bundle    *bundle.Bundle
+	Component *bundle.ComponentPackage
+}
+
+// AllComponents returns all the components in the BundleWrapper, with the
+// assumption that only one of Bundle or Component is filled out.
+func (bw *BundleWrapper) AllComponents() []*bundle.ComponentPackage {
+	if bw.Bundle != nil {
```
[the remark this answers is not in the document]
Comment (Kashomon) [reply pkg/commands/cmdlib/bundleio.go]: Yes, it's a union struct. I wish it were easier to specify these one-of situations. I think it's a little confusing, but I don't think it's too bad given that it's an internal only type to the commands. I could be wrong though

File: pkg/commands/cmdlib/bundleio.go [hunk 12]
```diff
@@ -29,88 +29,218 @@ import (
...
+
+// stdioReaderWriter is a source from stdin and sink to stdout.
+type stdioReaderWriter interface {
+	readBytes() ([]byte, error)
+	writeBytes([]byte) error
+}
+
+type realStdioReaderWriter struct{}
+
+func (r *realStdioReaderWriter) readBytes() ([]byte, error) {
```
[the remark this answers is not in the document]
Comment (Kashomon) [reply pkg/commands/cmdlib/bundleio.go]: I personally found it nice to inject the bigger struct-objects in the tests. Also the structs allow for nice fake-creation.

File: pkg/commands/cmdlib/bundleio.go [hunk 13]
```diff
@@ -29,88 +29,218 @@ import (
...
+		bytes = append(bytes, scanner.Bytes()...)
+		bytes = append(bytes, '\n')
+	}
+	if err := scanner.Err(); err != nil {
+		return nil, err
+	}
+	return bytes, nil
+}
+
+func (r *realStdioReaderWriter) writeBytes(b []byte) error {
```
[the remark this answers is not in the document]
Comment (Kashomon) [reply pkg/commands/cmdlib/bundleio.go]: I suppose so -- maybe this is overkill. But I like the way this fulfills the Writer contract this way. Seems like a style choice.

File: pkg/commands/cmdlib/bundleio.go [hunk 14]
```diff
@@ -29,88 +29,218 @@ import (
...
 	var bytes []byte
 	var err error
-	if g.BundleFile != "" {
-		log.Infof("Reading bundle file %v", g.BundleFile)
-		bytes, err = rw.ReadFile(ctx, g.BundleFile)
+	inFmt := g.InputFormat
+
+	if g.InputFile != "" {
+		log.Infof("Reading input file %v", g.InputFile)
+		bytes, err = brw.rw.ReadFile(ctx, g.InputFile)
```
[the remark this answers is not in the document]
Comment (Kashomon) [reply pkg/commands/cmdlib/bundleio.go]: No yet. I wanted to future proof for when I needed to do HTTP requests to do inlining.

File: pkg/commands/cmdlib/bundleio.go [hunk 15]
```diff
@@ -29,88 +29,218 @@ import (
...
 		return fmt.Errorf("error writing contents: %v", err)
 	}
-	return writeContents(ctx, g.OutputFile, bytes, rw)
+	return brw.writeContents(ctx, g.OutputFile, bytes, brw.rw)
 }
 
 // writeContents writes some bytes to disk or stdout. if outPath is empty,
 // write to stdout instead.
-func writeContents(ctx context.Context, outPath string, bytes []byte, rw files.FileReaderWriter) error {
+func (brw *BundleReaderWriter) writeContents(ctx context.Context, outPath string, bytes []byte, rw files.FileReaderWriter) error {
```
[the remark this answers is not in the document]
Comment (Kashomon) [reply pkg/commands/cmdlib/bundleio.go]: outPath is a path to a specific file. If not specified, it dumps the text to stdout.

File: pkg/commands/find/images.go [hunk 16]
```diff
@@ -39,12 +40,14 @@ func findAction(ctx context.Context, cmd *cobra.Command, _ []string) {
 }
 
 func runFindImages(ctx context.Context, _ *options, rw files.FileReaderWriter, gopt *cmdlib.GlobalOptions) error {
```
[the remark this answers is not in the document]
Comment (Kashomon) [reply pkg/commands/find/images.go]: I personally like the consistency more than I mind duplication or unnecessary params. It's easy to verify everything has similar behavior this way.

File: pkg/commands/root.go [hunk 17]
```diff
@@ -30,26 +30,24 @@ import (
...
 	rootCmd.PersistentFlags().StringVarP(
-		&cmdlib.GlobalOptionsValues.BundleFile, "bundle-file", "f", "", "The path to a bundle file")
+		&cmdlib.GlobalOptionsValues.InputFile, "input-file", "f", "", "The path to an input file")
 
 	rootCmd.PersistentFlags().StringVarP(
-		&cmdlib.GlobalOptionsValues.InputFormat, "in-format", "", "yaml", "The input file format. One of either 'json' or 'yaml'")
+		&cmdlib.GlobalOptionsValues.InputFormat, "in-format", "", "yaml", "The input file format. One of either 'json' or 'yaml'. If an input-file is specified, it is inferred from the file extension.")
 
 	rootCmd.PersistentFlags().StringVarP(
 		&cmdlib.GlobalOptionsValues.OutputFile, "output-file", "o", "", "The path for any output file")
```
[the remark this answers is not in the document]
Comment (Kashomon) [reply pkg/commands/root.go]: I've personally found this to be useful, but maybe it's overkill to have output format.

I definitely like supporting input files. Output files I'm less sure about. /shrug

**Comment by Kashomon** [comment]
> From an api standpoint, I would prefer format not to be inferred based upon the file extension. It's non-obvious

I disagree about this comment: I think it's actually non-obvious to not infer the content type from the file extension.

File: pkg/commands/cmdlib/bundleio.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (Kashomon) [reply pkg/commands/cmdlib/bundleio.go]: Changed to just use stdout

File: pkg/commands/root.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (Kashomon) [reply pkg/commands/root.go]: changed to just use stdout.

**Pull request merged**